### PR TITLE
Add optix to experimental

### PR DESCRIPTION
**Pull request by mtthss**
Introducing `optix`: a composable gradient processing and optimization library.

It's objective is to support the composition of arbitrary sets of gradient transformations,
including sequential transformations (e.g. clip then rescale by rms), and parallel transformation (where multiple distinct optimizers share a subset of the variable to optimize).

Many popular optimizers can be implemented as one-liners, and, for convenience,
we additionally provide aliases for the most common ones.

### CODE REVIEW

**Comment by mtthss** [comment]
@googlebot I signed it
(I am a google employee so I registered as such)

File: jax/experimental/optix.py [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+    updates = tree_multimap(lambda g: step_size_fn(state.count) * g, updates)
+    return updates, ScaleByScheduleState(count=state.count + 1)
+
+  return init_fn, update_fn
+
+
+### Utilities for building and using custom optimizers. ###
+
+
+def chainer(*args):
```
Comment (jekbradbury) [inline jax/experimental/optix.py]: I feel like this is maybe better as `chain`?

File: jax/experimental/optix.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+  Returns:
+    An (init_fn, update_fn) tuple.
+  """
+
+  def init_fn(params):
+    return TraceState(trace=tree_multimap(jnp.zeros_like, params))
+
+  def update_fn(updates, state):
+    f = lambda g, t: g + decay * t
+    update_trace = tree_multimap(f, updates, state.trace)
```
Comment (jekbradbury) [inline jax/experimental/optix.py]: Almost every optimizer `update_fn` (all of them except `clip_by_global_norm`) performs `tree_multimap` over an inner per-parameter update. Is that a pattern that you can extract out like `@optimizer` in optimizers.py or is there a reason why that doesn't work?

File: jax/experimental/optix.py [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+    updates: a tree of updates, the tree structure and the shape of the leaf
+    nodes must match that of `params`.
+
+  Returns:
+    Updated parameters, with same structure and shape as `params`.
+  """
+  return tree_multimap(lambda p, u: p + u, params, updates)
+
+
+### Aliases for popular optimizers. ###
```
Comment (jekbradbury) [inline jax/experimental/optix.py]: Are there one or two (maybe less popular) optimizers that you can add as examples of the benefits of composability? For instance, perhaps `nadam` from https://openreview.net/pdf?id=OM0jvwB8jIp57ZJjtNEZ or LARS can be expressed using `chainer` and the existing primitives?

File: jax/experimental/optix.py (hunk 2 above)
Comment (mtthss) [reply jax/experimental/optix.py]: I purposefully diverged from `optimizers.py` in this aspects for two reasons:
1) there are several gradient transformations of interest that require to consider the entire gradient and cannot be computer variable by variable. The `clip_by_global_norm` transformation included here is one example, but others could be PopArt, KFac, etc...
2) it really is just a handful of characters that are saved, and at the cost of introducing an additional level of indirection and a reduction in flexibility.

File: jax/experimental/optix.py (hunk 3 above)
Comment (mtthss) [reply jax/experimental/optix.py]: As an example I added a `noisy_sgd` as from the paper: https://arxiv.org/abs/1511.06807.

Crucially though, the composable nature of `optix` would allow the user to build in a one liner also a `noisy_adam` or `noisy_rmsprop`, or more generally combine the idea from the paper with any optimizer of his/her choice.

As a comparison to `TF` or `jax/experimental/optimizers.py`, there the user could only add noise to the gradient before applying the adam/rmsprop rescaling, because the adam/rmsprop would immediately apply the updates and the user could not insert itself before the update without rewriting the entire optimizer. 

Instead in optix the user may experiment with adding the noise before or after the adam/rmsprop rescaling (and I suspect the latter, very complicated without optix, will actually be better).

File: jax/experimental/optix.py (hunk 1 above)
Comment (mtthss) [reply jax/experimental/optix.py]: Done
